File: cmd/cwa/testdata/src/cwa/resource.rs
use raw;
use std::io;
use std::io::{Read, Write};

pub struct Resource {
    handle: i32
}

impl Resource {
    pub fn open(url: &str) -> Option<Resource> {
        let url = url.as_bytes();
        if url.len() == 0 {
            return None;
        }

        let handle = unsafe {
            raw::resource_open(&url[0], url.len())
        };
        if handle < 0 {
            None
        } else {
            Some(Resource {
                handle: handle
            })
        }
    }

    pub unsafe fn from_raw(handle: i32) -> Resource {
        // TODO: Deal with invalid handles (< 0)
        Resource {
            handle: handle
        }
    }
}

impl Drop for Resource {
    fn drop(&mut self) {
        unsafe {
            raw::resource_close(self.handle);
        }
    }
}

impl Read for Resource {
    fn read(&mut self, out: &mut [u8]) -> io::Result<usize> {
        let len = out.len();

        if len == 0 {
            return Ok(0);
        }

        let ret = unsafe { raw::resource_read(
            self.handle,
            &mut out[0],
            len
        ) };

        if ret < 0 {
            Err(io::Error::from(io::ErrorKind::Other))
        } else {
            Ok(ret as usize)
        }
    }
}

impl Write for Resource {
    fn write(&mut self, data: &[u8]) -> io::Result<usize> {
        let len = data.len();

        if len == 0 {
            return Ok(0);
        }

        let ret = unsafe { raw::resource_write(
            self.handle,
            &data[0],
            len
        ) };
        if ret < 0 {
            Err(io::Error::from(io::ErrorKind::Other))
        } else {
            Ok(ret as usize)
        }
    }

    fn flush(&mut self) -> io::Result<()> {
        // TODO
        Ok(())
    }
}

File: zig/src/olin/resource.zig
const errs = @import("./error.zig");
const OlinError = errs.OlinError;

extern fn resource_open(data: [*]const u8, len: usize) i32;
extern fn resource_read(fd: i32, data: [*]const u8, len: usize) i32;
extern fn resource_write(fd: i32, data: [*]const u8, len: usize) i32;
extern fn resource_close(fd: i32) void;
extern fn resource_flush(fd: i32) i32;

extern fn io_get_stdin() i32;
extern fn io_get_stdout() i32;
extern fn io_get_stderr() i32;

fn fd_check(fd: i32) errs.OlinError!Resource {
    if(errs.parse(fd)) |fd_for_handle| {
        return Resource {
            .fd = fd_for_handle,
        };
    } else |err| {
        return err;
    }
}

pub const open = Resource.open;
pub const stdin = Resource.stdin;
pub const stdout = Resource.stdout;
pub const stderr = Resource.stderr;

pub const Resource = struct {
    fd: i32,

    pub fn open(url: []const u8) !Resource {
        const fd = resource_open(url.ptr, url.len);
        return fd_check(fd);
    }

    pub fn stdin() !Resource{
        const fd = io_get_stdin();
        return fd_check(fd);
    }

    pub fn stdout() !Resource{
        const fd = io_get_stdout();
        return fd_check(fd);
    }

    pub fn stderr() !Resource{
        const fd = io_get_stderr();
        return fd_check(fd);
    }

    pub fn write(self: Resource, data: [*]const u8, len: usize) OlinError!i32 {
        const n = resource_write(self.fd, data, len);

        if (errs.parse(n)) |nresp| {
            return nresp;
        } else |err| {
            return err;
        }
    }

    pub fn read(self: Resource, data: [*]u8, len: usize) OlinError!i32 {
        const n = resource_read(self.fd, data, len);

        if (errs.parse(n)) |nresp| {
            return nresp;
        } else |err| {
            return err;
        }
    }

    pub fn close(self: Resource) void {
        resource_close(self.fd);
        return;
    }

    pub fn flush(self: Resource) OlinError!void {
        const n = resource_flush(self.fd);

        if (errs.parse(n)) {} else |err| {
            return err;
        }
    }
};

